samples: sensor: access_trig: add double tap detection.

pcurt opened the pull request:
The LIS2DW12 is an advanced 3-axis linear accelerometer. 
The driver exists in Zephyr but it is not very easy to configure it and use with trigger detection.

So I added a samples that configure the accelerometer to detect a double tap, the example is running on a **stm32f3_disco** board with  a **steval-mki190v1** adapter board

avisconti commented on samples/sensor/lis2dw12/README.rst:
```diff
@@ -0,0 +1,52 @@
+Overview
+********
+
+By defautlt this example uses the trigger mode with a double tap detection.
```
correct the typo ("default")

pcurt replied:
Fixed, thanks

avisconti commented on samples/sensor/lis2dw12/src/main.c:
```diff
@@ -0,0 +1,83 @@
+		fetch_and_display(sensor);
+		k_sleep(K_MSEC(2000));
+	}
+#endif /* CONFIG_LIS2DW12_TRIGGER */
```
I would keep a single ```while(true)``` block, with ```fetch_and_display()``` inside a #ifdef. The configuration of trigger under another ifdef above the while block.

pcurt replied:
If I do this I need two `#ifdef CONFIG_LIS2DW12_TRIGGER` conditions:
- one around the sensor_trigger_set bloc
- one around the` fetch_and_display(sensor)`
To keep a good code readability I prefer keep only one.

What do you think ?

avisconti replied:
Well, I think that maybe at the end it's just a matter of preference. I've nothing against it, just a matter of preference.

avisconti commented on samples/sensor/lis2dw12/src/main.c:
```diff
@@ -0,0 +1,83 @@
+	} else {
+		printf("#%u @ %u ms: x %f , y %f , z %f, t %f\n", count, k_uptime_get_32(),
+		       sensor_value_to_double(&accel[0]), sensor_value_to_double(&accel[1]),
+		       sensor_value_to_double(&accel[2]), sensor_value_to_double(&temperature));
```
sensor_balue_to_double() returns a double. Is it ok printing a float (%f)?

pcurt replied:
Yes, printf function automatically converts float in double, so it's ok here.

avisconti replied:
ok

avisconti reviewed (changes requested):
Just few comments.
Also note that you have a typo in the commit message ("aslo" instead of "also")

pcurt commented:
> Just few comments. Also note that you have a typo in the commit message ("aslo" instead of "also")

Fixed, thanks

MaureenHelm reviewed (changes requested):
We're trying to get away from driver-specific sensor samples. There are other sensor drivers that support the same  `SENSOR_TRIG_DOUBLE_TAP` and `SENSOR_CHAN_ACCEL_XYZ` channels, so if you'd like to add a sample demonstrating this functionality please make it independent of the specific driver implementation.

pcurt commented:
> We're trying to get away from driver-specific sensor samples. There are other sensor drivers that support the same `SENSOR_TRIG_DOUBLE_TAP` and `SENSOR_CHAN_ACCEL_XYZ` channels, so if you'd like to add a sample demonstrating this functionality please make it independent of the specific driver implementation.

I also think that it is better to have generic samples for multiple hardware. I can integrate the  `SENSOR_TRIG_DOUBLE_TAP` in the `accel_trig` example (https://docs.zephyrproject.org/latest/samples/sensor/accel_trig/README.html)
What do you think about it ?

pcurt committed:
samples: sensor: accel_trig: add sample tap detection feature.

Add the tap detection trigger to the existing example.
For this example it uses lis2dw12 accelerometer to detect a double
tap. This feature can be enabled with CONFIG_SAMPLE_TAP_DETECTION option.

Signed-off-by: Pierrick Curt <pierrickcurt@gmail.com>

pcurt commented:
@MaureenHelm I have integrated the tap detection in `accel_trig` example

pcurt commented:
This is the console output with TAP detection enabled
```
TAP detected
     lis2dw12@19 [m/s^2]:    (   -1.899901,   -12.550355,    -2.742174)
TAP detected
     lis2dw12@19 [m/s^2]:    (   12.349357,   -18.125630,     6.015556)
TAP detected
     lis2dw12@19 [m/s^2]:    (  -11.385050,    -7.274181,    -9.229117)
TAP detected
     lis2dw12@19 [m/s^2]:    (    9.214760,    -9.286545,     2.311466)
TAP detected
     lis2dw12@19 [m/s^2]:    (   10.090533,   -17.391034,    12.320643)
TAP detected
     lis2dw12@19 [m/s^2]:    (   -0.478564,     2.390429,    15.876378)
TAP detected
     lis2dw12@19 [m/s^2]:    (   -5.668596,   -13.138989,     0.741775)
TAP detected
     lis2dw12@19 [m/s^2]:    (   -2.385644,   -10.559526,     9.899107)
TAP detected
     lis2dw12@19 [m/s^2]:    (    7.537391,    -8.551948,    16.740187)
```

yperess reviewed (changes requested):
Can you also update the README file to reflect the feature?

pcurt commented on samples/sensor/accel_trig/src/main.c:
```diff
@@ -41,10 +55,19 @@ int main(void)
 		return 0;
 	}
 
+#if DT_NODE_HAS_PROP(DT_ALIAS(accel0), tap_mode)
```
You are right, this is _generic_ with sensor implementing the tap_mode, so only ST sensors.
But the others sensors doesn't have a tap configuration read from the device tree.
The other way to have a full generic sample would be to manage the trigger selection with a define in prj.conf. But I think that is not a very beautiful solution. What do you think ?

MaureenHelm replied:
Agree with @kartben . I think you could just add a Kconfig option to the sample to enable/disable using the tap trigger.

pcurt replied:
I have added the Kconfig

MaureenHelm reviewed (commented):
> So I added a samples that configure the accelerometer to detect a double tap, the example is running on a stm32f3_disco board with a steval-mki190v1 adapter board

It would be nice to add shield board support for the steval-mki190v1 instead of adding a devicetree overlay in the sample.
https://docs.zephyrproject.org/latest/hardware/porting/shields.html

pcurt commented:
> > So I added a samples that configure the accelerometer to detect a double tap, the example is running on a stm32f3_disco board with a steval-mki190v1 adapter board
> 
> It would be nice to add shield board support for the steval-mki190v1 instead of adding a devicetree overlay in the sample. https://docs.zephyrproject.org/latest/hardware/porting/shields.html

Good idea the shield,  when I have looked on all supported shield, I have found that the `x_nucleo_iks01a3` integrates `lis2dw12` so it is be better to use it for the example.
I'm on it

avisconti commented:
> > > So I added a samples that configure the accelerometer to detect a double tap, the example is running on a stm32f3_disco board with a steval-mki190v1 adapter board
> > 
> > 
> > It would be nice to add shield board support for the steval-mki190v1 instead of adding a devicetree overlay in the sample. https://docs.zephyrproject.org/latest/hardware/porting/shields.html
> 
> Good idea the shield, when I have looked on all supported shield, I have found that the `x_nucleo_iks01a3` integrates `lis2dw12` so it is be better to use it for the example. I'm on it

I think you can add both overlays.

IAmAnkur commented on samples/sensor/accel_trig/README.rst:
```diff
@@ -7,6 +7,10 @@ Overview
 ********
 
 This sample application demonstrates how to use 3-Axis accelerometers with triggers.
+By defaut it uses a data ready trigger to read the accelerometer data and print it to the console.
```
small typo "By default"

pcurt replied:
Thanks

pcurt commented:
> > > > So I added a samples that configure the accelerometer to detect a double tap, the example is running on a stm32f3_disco board with a steval-mki190v1 adapter board
> > > 
> > > 
> > > It would be nice to add shield board support for the steval-mki190v1 instead of adding a devicetree overlay in the sample. https://docs.zephyrproject.org/latest/hardware/porting/shields.html
> > 
> > 
> > Good idea the shield, when I have looked on all supported shield, I have found that the `x_nucleo_iks01a3` integrates `lis2dw12` so it is be better to use it for the example. I'm on it
> 
> I think you can add both overlays.

OK

MaureenHelm commented on samples/sensor/accel_trig/boards/nrf52dk_nrf52832.overlay:
```diff
@@ -0,0 +1,29 @@
+#include <zephyr/dt-bindings/sensor/lis2dw12.h>
+
+/ {
+	aliases {
+		accel0 = &lis2dw12_19_x_nucleo_iks01a3;
+	};
+};
+
+&arduino_i2c {
+
+	lis2dw12_19_x_nucleo_iks01a3: lis2dw12@19 {
+		status = "okay";
+		compatible = "st,lis2dw12";
+		reg = <0x19>;
+		odr = <400>;
+		tap-mode = <LIS2DW12_DT_SINGLE_DOUBLE_TAP>;
+		tap-threshold = <12>, <12>, <12>;
+		power-mode = <LIS2DW12_DT_HP_MODE>;
+		tap-shock = <0x03>;
+		tap-quiet = <0x03>;
+		tap-latency = <0x03>;
+		irq-gpios =  <&arduino_header 3 GPIO_ACTIVE_HIGH>; /* A3 */
+	};
```
This can be simplified to
```suggestion
&lis2dw12_19_x_nucleo_iks01a3 {
	status = "okay";
	odr = <400>;
	tap-mode = <LIS2DW12_DT_SINGLE_DOUBLE_TAP>;
	tap-threshold = <12>, <12>, <12>;
	power-mode = <LIS2DW12_DT_HP_MODE>;
	tap-shock = <0x03>;
	tap-quiet = <0x03>;
	tap-latency = <0x03>;
};
```
And with some cmake magic, you could move it to a common overlay that gets conditionally applied when SHIELD=x_nucleo_iks01a3, instead of duplicating it for nrf52dk_nrf52832 and stm32f3_disco.

pcurt replied:
Yes, thanks for the suggestion

MaureenHelm commented on samples/sensor/accel_trig/Kconfig:
```diff
@@ -0,0 +1,13 @@
+
+mainmenu "Accelerometer trigger sample application"
+
+config SAMPLE_TAP_DETECTION
+bool "Set tap detection as trigger in the sample"
+default n
```
```suggestion
config SAMPLE_TAP_DETECTION
	bool "Set tap detection as trigger in the sample"
```

pcurt replied:
done

MaureenHelm reviewed (commented):
Overall direction looks good, just a couple of nits.

aescolar commented:
@pcurt note this PR introduced a bug in main:
https://github.com/zephyrproject-rtos/zephyr/actions/runs/13346970336/job/37278635760
`devicetree error: samples/sensor/accel_trig/boards/nrf52dk_nrf52832.overlay:8 (column 1): parse error: undefined node label 'lis2dw12_19_x_nucleo_iks01a3'`
It seems the new overlay introduced for nrf52dk_nrf52832 in this PR needs a shield (x_nucleo_iks01a3 ?).

pcurt commented:
> @pcurt note this PR introduced a bug in main: https://github.com/zephyrproject-rtos/zephyr/actions/runs/13346970336/job/37278635760 `devicetree error: samples/sensor/accel_trig/boards/nrf52dk_nrf52832.overlay:8 (column 1): parse error: undefined node label 'lis2dw12_19_x_nucleo_iks01a3'` It seems the new overlay introduced for nrf52dk_nrf52832 in this PR needs a shield (x_nucleo_iks01a3 ?). The issue can be reproduced with `twister -p nrf52dk/nrf52832 -T samples/sensor/accel_trig`

Sorry, I'll take care of suggesting a correction.
